Migrate to MV3 (Chrome, Firefox, Edge)

Sneezry opened the pull request:
Fixes #565

- [x] Move to `chrome.storage.session` to cache password
- [x] Move to `chrome.scripting` to inject scan QR script
- [x] Replace `content_security_policy` with `content_security_policy.extension_pages` or `content_security_policy.sandbox` as appropriate.
- [x] Replace `setTimeout` with `chrome.alarms` API for auto-lock
- [x] Replace `XMLHttpRequest` with `fetch` in backend agent
- [x] Migrate LocalStorage to `chrome.storage.local`
- [x] <del>Migrate `Image` and `canvas` to `OffscreenCanvas`(?) for service worker use</del> Move QR decode to content script
- [ ] E2E test
  + [ ] Chrome 
	  + [x] Can create new accounts
	  + [x] Can create new accounts by scanning QR codes
	  + [x] Can import existing accounts
	  + [x] Can migrate data with existing installed MV2 extension
	  + [x] Can set password
	  + [x] Can click to copy
	  + [x] Can click to auto fill
	  + [ ] Can auto backup to 3rd party storage
	  + [x] Auto-lock works as expected
	  + [x] Force lock works as expected
	  + [x] Context menu works as expected
	  + [x] Policy configuration works as expected
	  + [x] Storage area migration works as expected
	  + [ ] TBD
  + [ ] Firefox
	  + [x] Can create new accounts
	  + [x] Can create new accounts by scanning QR codes
	  + [x] Can import existing accounts
	  + [x] Can migrate data with existing installed MV2 extension
	  + [x] Can set password
	  + [x] Can click to copy
	  + [x] Can click to auto fill
	  + [x] Can auto backup to 3rd party storage
	  + [x] Auto-lock works as expected
	  + [x] Force lock works as expected
	  + [ ] ~Context menu works as expected~
	  + [ ] Policy configuration works as expected
	  + [x] Storage area migration works as expected
	  + [ ] TBD
  + [ ] Edge
	  + [x] Can create new accounts
	  + [x] Can create new accounts by scanning QR codes
	  + [x] Can import existing accounts
	  + [x] Can migrate data with existing installed MV2 extension
	  + [x] Can set password
	  + [x] Can click to copy
	  + [x] Can click to auto fill
	  + [ ] Can auto backup to 3rd party storage
	  + [x] Auto-lock works as expected
	  + [x] Force lock works as expected
	  + [x] Context menu works as expected
	  + [ ] Policy configuration works as expected
	  + [x] Storage area migration works as expected
	  + [ ] TBD

## Known issues

- ~Backup export failed when setting a password~
- ~Browser sync is disabled by default, need to double confirm if it is as expected~

Sneezry committed:
Migrate to MV3

Sneezry committed:
Migrate to MV3

Sneezry committed:
fix entry type

Sneezry committed:
fix entry type

Sneezry committed:
fix default storage area

Sneezry committed:
fix default storage area

Sneezry commented:
Some scenario test is still on the way, but the code should be ready for review.

Sneezry commented:
This PR only contains Chrome parts. Firefox and Edge MV3 migration should be in other PRs.

Sneezry commented:
@rebornix Chrome, Firefox, and Edge are starting to migrate to MV3, MV2 will stop working in the next few months. There are many breaking changes in MV3, Safari may also have some updates. Once this PR has been merged, the Safari build may be broken.

mymindstorm commented:
Just FYI, I might not be able to get to this until next Wednesday.

mymindstorm commented:
I'm getting 

```
/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:133
                if(isError) throw e;
                            ^

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:71:19)
    at Object.createHash (node:crypto:133:10)
    at module.exports (/home/mymindstorm/Projects/Authenticator-1/node_modules/webpack/lib/util/createHash.js:135:53)
    at NormalModule._initBuildHash (/home/mymindstorm/Projects/Authenticator-1/node_modules/webpack/lib/NormalModule.js:417:16)
    at handleParseError (/home/mymindstorm/Projects/Authenticator-1/node_modules/webpack/lib/NormalModule.js:471:10)
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/webpack/lib/NormalModule.js:503:5
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/webpack/lib/NormalModule.js:358:12
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:373:3
    at iterateNormalLoaders (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:214:10)
    at iterateNormalLoaders (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:221:10)
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:236:3
    at context.callback (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:111:13)
    at makeSourceMapAndFinish (/home/mymindstorm/Projects/Authenticator-1/node_modules/ts-loader/dist/index.js:90:5)
    at successLoader (/home/mymindstorm/Projects/Authenticator-1/node_modules/ts-loader/dist/index.js:68:9)
    at Object.loader (/home/mymindstorm/Projects/Authenticator-1/node_modules/ts-loader/dist/index.js:22:12)
    at LOADER_EXECUTION (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:119:14)
    at runSyncOrAsync (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:120:4)
    at iterateNormalLoaders (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:232:2)
    at Array.<anonymous> (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:205:4)
    at Storage.finished (/home/mymindstorm/Projects/Authenticator-1/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:55:16)
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:91:9
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/graceful-fs/graceful-fs.js:123:16
    at FSReqCallback.readFileAfterClose [as oncomplete] (node:internal/fs/read_file_context:68:3) {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}

Node.js v18.12.1
```

when building. It goes away with `export NODE_OPTIONS=--openssl-legacy-provider`.

 I think this might be fixed if we use a newer version of loader-runner (seems to be a dependency of webpack). Would that be easy to do here?


---

The service worker isn't registering properly on my Chrome: `Service worker registration failed. Status code: 15`.

![image](https://user-images.githubusercontent.com/27789806/220240646-6f2ce98f-9a5d-457f-8cb4-957bf996180c.png)

---

Keyboard shortcuts seem to not be working due to manifest issue:

![image](https://user-images.githubusercontent.com/27789806/220240804-ab8c773e-1eaf-4e83-962f-7038c6d4b030.png)

Sneezry commented:
Can you load this build locally?
[chrome.zip](https://github.com/Authenticator-Extension/Authenticator/files/10789323/chrome.zip)

mymindstorm commented:
That build has same issues with the service worker and keyboard commands. (Chromium Version 110.0.5481.77)

Sneezry commented:
Interesting... It works fine on my machine though...

![image](https://user-images.githubusercontent.com/1621293/220242490-8e0dbd82-9afa-483d-b62b-1e09b59de843.png)

![image](https://user-images.githubusercontent.com/1621293/220242531-f1ca9d29-bc67-4728-ac96-e99b04c89066.png)

![image](https://user-images.githubusercontent.com/1621293/220242564-d99fe754-55a7-483d-97e9-3d7b66cf1b47.png)

mymindstorm commented:
Interesting, I'll investigate further this weekend. What is your chrome version?


-------- Original message --------
From: Zhe Li ***@***.***>
Date: 2/20/23 9:47 PM (GMT-06:00)
To: Authenticator-Extension/Authenticator ***@***.***>
Cc: Brendan Early ***@***.***>, Review requested ***@***.***>
Subject: Re: [Authenticator-Extension/Authenticator] Migrate to Chrome MV3 (PR #1009)


Interesting... It works fine on my machine though...

[image]<https://user-images.githubusercontent.com/1621293/220242490-8e0dbd82-9afa-483d-b62b-1e09b59de843.png>

[image]<https://user-images.githubusercontent.com/1621293/220242531-f1ca9d29-bc67-4728-ac96-e99b04c89066.png>

[image]<https://user-images.githubusercontent.com/1621293/220242564-d99fe754-55a7-483d-97e9-3d7b66cf1b47.png>

—
Reply to this email directly, view it on GitHub<https://github.com/Authenticator-Extension/Authenticator/pull/1009#issuecomment-1437819555>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGUAT3TEG4BOBV3HDW2IS7DWYQ3DZANCNFSM6AAAAAAUM4FOEQ>.
You are receiving this because your review was requested.Message ID: ***@***.***>

mymindstorm commented:
oops, I see that you've attached it.

Sneezry commented:
I might find the root cause of the service work register issue. 

> Lots of API inside chrome are present only if your extension's manifest.json lists its associated permission name in "permissions" or in a special key
> https://stackoverflow.com/a/73789284

The contextMenu is in optional permission, which means `chrome.contextMenu` should be undefined before the permission granted.

We can move it to required permission since I see no security risk about that.

@mymindstorm do you have any concern?

mymindstorm commented:
Sorry for the lateness on this. I will get to it as soon as I can.

mymindstorm commented:
I agree, I don't see a security concern there.

mymindstorm committed:
fix mv3 migration errors

mymindstorm committed:
fix mv3 migration errors

mymindstorm commented:
I think I've resolved most of the issues I found. Please make sure to test on your end too to make sure I didn't miss anything.

mymindstorm committed:
further CI fixes

mymindstorm committed:
further CI fixes

mymindstorm commented:
I am leaving the tests broken, I ran out of time to fix them.

Sneezry commented:
> I think I've resolved most of the issues I found. Please make sure to test on your end too to make sure I didn't miss anything.

Awesome! It works w/o problems on my side.

Sneezry committed:
Fix disable backup policy not working issue (#1050)

Sneezry committed:
6.3.5

DNin01 commented on manifests/manifest-chrome.json:
```diff
@@ -41,32 +41,36 @@
+      "https://graph.microsoft.com/me/*",
+      "https://login.microsoftonline.com/common/oauth2/v2.0/token"
     ],
     "offline_enabled": true,
```
I can't find any information that suggests that `offline_enabled` is still supported by Manifest V3. Compare these pages:
https://developer.chrome.com/docs/extensions/mv2/manifest/
https://developer.chrome.com/docs/extensions/mv3/manifest/

[The page for this property on MDN Web Docs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/offline_enabled) also lists the supported manifest version as 2 only.

Sneezry replied:
Thanks for the information.

spaette committed:
typo

mymindstorm committed:
Merge pull request #1138 from vaerksted/dev

fix typo

mymindstorm committed:
Fix bad practice with argon2-browser

mymindstorm committed:
Fix bad practice with argon2-browser

mymindstorm committed:
Remove 'offline_enabled' from manifest

mymindstorm committed:
Remove 'offline_enabled' from manifest

mymindstorm committed:
update test runner and coverage for mv3

mymindstorm committed:
update test runner and coverage for mv3

mymindstorm committed:
don't use dev config by default

mymindstorm committed:
don't use dev config by default

mymindstorm commented:
I spent a few hours working on tests. They technically work again. Except for the fact that I had to disable the UI ones until I have more time to work out the kinks of MV3 disallowing all unsafe-eval and some weird webpack dependency issues. At least it is in an overall better state than it was previously.

vuittont60 committed:
fix typo (#1144)

Sneezry committed:
add import QR images/OTP URLs buttons to add account page

Sneezry committed:
Merge branch 'mv3' of github.com:Authenticator-Extension/Authenticator into mv3

Sneezry committed:
dump version

Sneezry committed:
update manifest

Sneezry committed:
fix ff csp

Sneezry commented:
Firefox should now be operational. However, end-to-end testing has not yet been completed.

WofWca commented:
Hey! So, the extension seems to be working alright on Manifest V3, it's only tests (i.e. non-production code) that are missing? Am I reading this correctly?

Sneezry commented:
> Hey! So, the extension seems to be working alright on Manifest V3, it's only tests (i.e. non-production code) that are missing? Am I reading this correctly?

The end-to-end test is not completed. If all tests pass, we can deliver it to the public.

mymindstorm committed:
remove artifact

mymindstorm committed:
rename test files

mymindstorm committed:
move syncTimeWithGoogle out of popup.ts

this prevents a dependency issue in the tests

mymindstorm committed:
fix the tests, remove code coverage

mv3 makes code cov w/ istanbul virtually impossible due to restrictions on unsafe-eval

mymindstorm commented:
Tests work again! We lost the ability for code coverage reporting, but that was half broken due to vue anyway.

![image](https://github.com/Authenticator-Extension/Authenticator/assets/27789806/927e5f15-e3a4-4239-88c5-fb910a1a193c)

mymindstorm committed:
remove testing code

mymindstorm commented on src/models/storage.ts:
```diff
@@ -48,28 +59,33 @@ export class BrowserStorage {
   }
 
   // TODO: promise this
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
remove these comments

mymindstorm commented on src/models/storage.ts:
```diff
@@ -48,28 +59,33 @@ export class BrowserStorage {
-  static async get(callback: (items: { [key: string]: any }) => void) {
+  static async get() {
     const storageLocation = await this.getStorageLocation();
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
remove this comment

mymindstorm reviewed (changes requested):
I really don't like how `LocalStorage` works. 

 - Putting it as a global everywhere makes it easy to misuse (e.g. forget to load/store data from/to `chrome.storage` again). This should all be hidden behind a class. It's too easy to make mistakes with the current way.
 - Using the global var as a sort-of cache introduces the possibility of race conditions
 - We should name is something more descriptive, like `userSettings`
 - We know exactly what should be in `LocalStorage`/`userSettings`. There should be an interface for it instead of `any`.

For example, something like:

```ts
// UserSettings.d.ts
declare type StorageLocation = "sync" | "local";

interface UserSettings {
    storageLocation?: StorageLocation;
}

// LocalUserSettings.ts
export class LocalUserSettings {
  static async getUserSettings(): Promise<UserSettings> {
    return (await chrome.storage.local.get("userSettings"))?.userSettings || {};
  }

  static async setUserSettings(userSettings: UserSettings) {
    let mergedSettings = await getUserSettings();
    Object.assign(mergedSettings, userSettings);
    await chrome.storage.local.set({ userSettings: mergedSettings });
  }
}
```

What are your thoughts on this?

mymindstorm commented:
Also, the migration for localStorage broke in Firefox. Will investigate when I have more time.

Sneezry commented:
> I really don't like how `LocalStorage` works.
> 
> * Putting it as a global everywhere makes it easy to misuse (e.g. forget to load/store data from/to `chrome.storage` again). This should all be hidden behind a class. It's too easy to make mistakes with the current way.
> * Using the global var as a sort-of cache introduces the possibility of race conditions
> * We should name is something more descriptive, like `userSettings`
> * We know exactly what should be in `LocalStorage`/`userSettings`. There should be an interface for it instead of `any`.
> 
> For example, something like:
> 
> ```ts
> // UserSettings.d.ts
> declare type StorageLocation = "sync" | "local";
> 
> interface UserSettings {
>     storageLocation?: StorageLocation;
> }
> 
> // LocalUserSettings.ts
> export class LocalUserSettings {
>   static async getUserSettings(): Promise<UserSettings> {
>     return (await chrome.storage.local.get("userSettings"))?.userSettings || {};
>   }
> 
>   static async setUserSettings(userSettings: UserSettings) {
>     let mergedSettings = await getUserSettings();
>     Object.assign(mergedSettings, userSettings);
>     await chrome.storage.local.set({ userSettings: mergedSettings });
>   }
> }
> ```
> 
> What are your thoughts on this?

That makes sense. I will update the code later.

mymindstorm commented:
Ok. I will be unavailable for the next week and a half, but can help out after that.

Sneezry committed:
refactor user settings (#1191)

Sneezry committed:
remove out-of-date eslint comments

Sneezry commented:
@mymindstorm I have updated the PR. Let's complete this PR before end of the next week (5/26).

Sneezry committed:
fix user settings migration issue

Sneezry committed:
fix user setting migration issue

Sneezry commented:
> Also, the migration for localStorage broke in Firefox. Will investigate when I have more time.

I didn't meet any error when migrating from MV2. Could you test the latest code to verify if the issue has been resolved?

Sneezry committed:
fix edge errors

mymindstorm commented:
Sure. I've been sick since last week so it might take me some time to get to.


-------- Original message --------
From: Zhe Li ***@***.***>
Date: 5/19/24 11:47 PM (GMT-06:00)
To: Authenticator-Extension/Authenticator ***@***.***>
Cc: Brendan Early ***@***.***>, Mention ***@***.***>
Subject: Re: [Authenticator-Extension/Authenticator] Migrate to MV3 (Chrome, Firefox, Edge) (PR #1009)


Also, the migration for localStorage broke in Firefox. Will investigate when I have more time.

I didn't meet any error when migrating from MV2. Could you test the latest code to verify if the issue has been resolved?

—
Reply to this email directly, view it on GitHub<https://github.com/Authenticator-Extension/Authenticator/pull/1009#issuecomment-2119659227>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGUAT3QA5A42QZRE66CI6GTZDF55BAVCNFSM6AAAAAAUM4FOESVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMJZGY2TSMRSG4>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Sneezry commented:
Oh, take care and recover soon!

Brendan Early ***@***.***>于2024年5月20日 周一20:46写道：

> Sure. I've been sick since last week so it might take me some time to get
> to.
>
>
> -------- Original message --------
> From: Zhe Li ***@***.***>
> Date: 5/19/24 11:47 PM (GMT-06:00)
> To: Authenticator-Extension/Authenticator ***@***.***>
> Cc: Brendan Early ***@***.***>, Mention ***@***.***>
> Subject: Re: [Authenticator-Extension/Authenticator] Migrate to MV3
> (Chrome, Firefox, Edge) (PR #1009)
>
>
> Also, the migration for localStorage broke in Firefox. Will investigate
> when I have more time.
>
> I didn't meet any error when migrating from MV2. Could you test the latest
> code to verify if the issue has been resolved?
>
> —
> Reply to this email directly, view it on GitHub<
> https://github.com/Authenticator-Extension/Authenticator/pull/1009#issuecomment-2119659227>,
> or unsubscribe<
> https://github.com/notifications/unsubscribe-auth/AGUAT3QA5A42QZRE66CI6GTZDF55BAVCNFSM6AAAAAAUM4FOESVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMJZGY2TSMRSG4>.
>
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Authenticator-Extension/Authenticator/pull/1009#issuecomment-2120392085>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAML2LK6LWI4FEUZLYDRO4TZDHWBJAVCNFSM6AAAAAAUM4FOESVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMRQGM4TEMBYGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Sneezry committed:
fix edge issues

Sneezry committed:
update firefox permissions

Sneezry committed:
remove all_urls permission since Firefox has supported activeTab

Sneezry commented:
Considering Google is required all extensions to migrate to MV3 by June, and leaving a week for CWS review, I will complete this PR by 5/26 anyway.

mymindstorm commented:
Just FYI, starting to work on this. Will probably push something within 3-6 hours.

mymindstorm committed:
fix firefox crash due to functions getting added to usersettings object

mymindstorm reviewed (approved):
From what I've seen it appears to work fine, outside of that crash I found. Although, I want to start automating E2E tests at some point.
